MQTT V5 support

icraggs opened the pull request:
These are my initial changes for MQTT V5 support.  They allow V5 packets to be sent and received, as shown in the test_mqttv5.py.   I'm submitting the changes at this stage for any comments or suggestions before I continue.  

Remaining work:

- V5 functionality: AUTH packets (probably for later), handling incoming DISCONNECT packet, session expiry (clean start vs clean session), server receive maximum (flow control)
- doc
- tests
- examples

icraggs commented:
OK, I've made this work in Python2.7 as well as 3.

PierreF commented on src/paho/mqtt/properties.py:
```diff
@@ -0,0 +1,370 @@
+  zz = buf.find("\x00") # look for null in the UTF string
+  if zz != -1:
+    raise MalformedPacket("[MQTT-1.5.4-2] Null found in UTF data "+buf)
+  for c in range (0xD800, 0xDFFF):
```
I fear about the cost of this. We are iterating over a rather large range (2047 item) times the buffer length which could also be rather long.
I would said the if we are iterating over the buffer anyway, it may be cheaper to do:
```
for c in buf:
  if c >= 0xD800 and c <= 0xDFFF:
    raise MalformedPacket
```

icraggs replied:
We could have a 'strict' option in the library which turns such checks on or off?  Not sure what the default would be.

I would have thought the payload would potentially be the biggest piece of data, but yes, the UTF-8 strings could be large too.

icraggs replied:
Fixed as suggested.

PierreF commented on src/paho/mqtt/properties.py:
```diff
@@ -0,0 +1,370 @@
+      # check that this attribute applies to the packet type
+      if self.packetType not in self.properties[self.getIdentFromName(name)][1]:
+        raise MQTTException("Property %s does not apply to packet type %s"
+            % (name, Packets.Names[self.packetType]) )
```
"Packets" variable does not exists

icraggs replied:
True.  I can add the Names array into the PacketTypes class.

icraggs replied:
Fixed.

PierreF commented:
About the test, I think it would be better to use same kind of tests as already present.
There is already two kind of test: "test" folder which use old tests and "tests" folder which use new testing infrastructure (py.test) but we hadn't moved all tests yet.

icraggs commented on src/paho/mqtt/properties.py:
```diff
@@ -0,0 +1,370 @@
+  buffer += [length // 65536]
+  length %= 65536
+  buffer += [length // 256, length % 256]
+  return bytearray(buffer)
```
No - I just never used the struct package, so my first contribution is just as is, with minimal changes, showing it works before any refactoring.  The test program shows it works.  We can replace code with struct.pack where appropriate.

icraggs replied:
Changed to use struct package.

icraggs commented on src/paho/mqtt/client.py:
```diff
@@ -2387,12 +2442,38 @@ def _send_connect(self, keepalive, clean_session):
+                    remaining_length += len(packed_props)
+                    packet += packed_props
+
+        self._pack_remaining_length(packet, remaining_length)
```
It shouldn't.  It's just increasing the remaining length value to include that of the properties too.  The properties are serialized first to determine their length, then added afterwards.

icraggs commented on src/paho/mqtt/client.py:
```diff
@@ -1288,21 +1298,36 @@ def subscribe(self, topic, qos=0):
 
         if isinstance(topic, tuple):
             topic, qos = topic
+            if self._protocol == MQTTv5 and not isinstance(qos, SubscribeOptions):
```
If you want to know more about MQTT V5, and the differences between it and 3.1.1:

- http://modelbasedtesting.co.uk/2018/04/09/a-story-of-mqtt-5-0/
- http://modelbasedtesting.co.uk/talks-ive-given/
- https://www.slideshare.net/IanCraggs/where-next-for-mqtt

icraggs replied:
This check is in the case where the 'topic' argument is a tuple.  For MQTT v3, the tuple is

(topic, QoS)

whereas for MQTT V5, the tuple is 

(topic, SubscribeOptions)

AtosNicoS commented:
Is there any update on this? :)

icraggs commented:
Sorry, I'm working on the Java client release (https://github.com/eclipse/paho.mqtt.java/milestone/2) as my top priority.  I'll get back to this as soon as I can.   I'm open to any offers of help.

icraggs commented:
Oh - obviously I should make the tests work!  And there are some other comments I haven't yet addressed.

icraggs commented on src/paho/mqtt/client.py:
```diff
@@ -2347,6 +2383,20 @@ def _send_connect(self, keepalive, clean_session):
+                packed_connect_properties = b'\x00'
+            else:
+                packed_connect_properties = self._connect_properties.pack()
+                print("packet connect properties", len(packed_connect_properties))
```
Removed.

icraggs commented on src/paho/mqtt/client.py:
```diff
@@ -1232,16 +1242,16 @@ def enable_bridge_mode(self):
+        return self._send_disconnect(reasoncode, properties)
 
-    def subscribe(self, topic, qos=0):
+    def subscribe(self, topic, qos=0, options=None, properties=None):
```
Option 3 done.

icraggs commented:
I've addressed all the points that Pierre made.  I've started on the tests - got about a third of the way through those.  The remaining work is finishing off the test suite and the documentation.  For anyone who wants to try it out, it's basically working.

icraggs commented:
@PierreF I've now finished the test suite as I intend for now.  The only remaining work is the documentation.  If it's possible to take a look before I update the doc to see if you're happy or suggest other changes, that would be really useful.

icraggs commented on src/paho/mqtt/client.py:
```diff
@@ -1269,16 +1279,16 @@ def enable_bridge_mode(self):
         self._client_mode = MQTT_BRIDGE
 
-    def disconnect(self):
+    def disconnect(self, reasoncode=None, properties=None):
```
I hadn't really thought about it.  There are some reason codes which I think are, or could be, an application decision, such as:

- disconnect with will message: 'the client wished to disconnect but requires that the server also publishes the will message'
- message rate too high
- administrative action
- quota exceeded

PierreF replied:
Agree. So let's keep reason-code here.

icraggs replied:
ok

icraggs commented on src/paho/mqtt/client.py:
```diff
@@ -1220,7 +1230,7 @@ def publish(self, topic, payload=None, qos=0, retain=False):
 
                     rc = self._send_publish(message.mid, topic, message.payload, message.qos, message.retain,
-                                            message.dup)
+                                            message.dup, MQTTMessageInfo(local_mid), message.properties)
```
Yes, looks like it.  I'm not sure why I did that.

icraggs replied:
Fixed.

icraggs commented on src/paho/mqtt/client.py:
```diff
@@ -490,8 +498,8 @@ def on_connect(client, userdata, flags, rc):
 
-    def __init__(self, client_id="", clean_session=True, userdata=None,
-                 protocol=MQTTv311, transport="tcp"):
+    def __init__(self, client_id="", clean_session=True, clean_start=True,
```
I'll fix.

icraggs replied:
Done.

icraggs commented on src/paho/mqtt/client.py:
```diff
@@ -523,11 +531,25 @@ def __init__(self, client_id="", clean_session=True, userdata=None,
-            raise ValueError('A client id must be provided if clean session is False.')
+
+        if protocol == MQTTv5:
+            if not clean_session:
```
The names 'clean session' and 'clean start' are used in the MQTT specifications for 3.1.1 and 5.0, so I thought it was the right thing to do to use those names for the parameters.  

We could have one parameter called clean_session_or_start, but that would cause some backward compatibility problems.  Is that what you were thinking?

PierreF replied:
I think of two options:

* Either keep clean_session and clean_start options but just change the default from True to None. This will also to raise the error as soon as user specify either True or False
* Or make clean_session an alias of clean_start... we may achieve this using **kwargs. But I fear this solution is too complicated for the gain.

icraggs replied:
I agree that having clean_session and clean_start as synonyms would be ideal, but also complicated to implement.

Actually this brings up another consideration.  I was surprised to see the clean_session option on the constructor (I haven't used this Python client very much), rather than the connect method.

In MQTT 5.0, the session cleanup is governed by the session expiry property, which defaults to 0.  This means that if you want to clean up at the start of the session, but be able to reconnect and continue if there are communication breaks until your work is done, you can do it in one go:

connect clean_start=true, session_expiry=99999 (value in seconds)

disconnect also has a session expiry property defaulting to 0, so when you issue a normal disconnect, then the session is cleaned up immediately.

To me, it makes sense to have the clean_start and session_expiry property on the same method.  One way of making this work would be to have the clean_start parameter on connect rather than the constructor, along with the connect properties.  Then clean_session on the constructor would not conflict with clean_start, and would not be allowed with MQTT 5.0.

icraggs commented on src/paho/mqtt/client.py:
```diff
@@ -1324,22 +1334,43 @@ def subscribe(self, topic, qos=0):
-                topic_qos_list.append((t.encode('utf-8'), q))
+            if self._protocol == MQTTv5:
+                for t, o in topic:
+                    if not isinstance(o, SubscribeOptions):
```
I'm happy to do that.  I think it's probably the best approach

icraggs replied:
Fixed

icraggs commented:
I ran the mqttv5 test program against Mosquitto 1.6.2 earlier.  Many things/tests worked, some failed.  Some failures were caused by configuration changes were in Mosquitto.  Some seemed to show Mosquitto bugs, and there was one crash in Mosquitto, in the will delay test.

icraggs commented:
@PierreF Any other tasks/changes (apart from doc completion) which you'd like me to do/make?  I did run the tests against Mosquitto and found some issues, some of which are definitely in Mosquitto.

PierreF commented:
Apart from running test that may need some improvement (if I run tox on my system, it will fail because the testing/interoperability broker is not running), I don't have more changes to do.

Are you going to make documentation change in this PR ?

About running the test, I'm not sure what is the best solution... I think to some kind of conditional test, like "tox --enable-interoperability-tests" which would be used by travis or someone which run the interop broker. But I don't known what the proper way to do this.
If you had time to look at this, it would be great. If not, I'll fix it later (I think using an environment variable and pytest skipif may do the job).

icraggs commented:
@PierreF Thanks.  Yes, it is my intention to add the docs in this PR, I just wanted to get your feedback before doing that.  Did you get chance to try these changes out with Mosquitto?

I'll look into getting the tox tests running without having to run the interop broker separately - it hadn't occurred to me.

PierreF commented:
No, I didn't tested them with Mosquitto. But the test_mqtt5 with interop broker seems to have good coverage of MQTT 5. That still on my todo, but I don't think it's mandatory before merging this PR.

ralight commented:
I've run test_mqtt5 against Mosquitto and it all seems fine, apart from a few issues with Mosquitto which have been highlighted and mostly fixed.

Having said that, I'm seeing some failures in `test_client_topic_alias` and `test_session_expiry` when I run the whole suite but that run fine on their own.

icraggs commented:
@ralight there are likely still some timing foibles in the tests, which I think are down to the way the tests are written rather than the client or server behaviour errors.  I found stopping and restarting the background thread loop often quite unwieldy for the tests I wanted to write.  Maybe there's a better way.

icraggs commented:
And @ralight, while you're here :-) I wonder if you have a view on this:

In MQTT 5.0, the session cleanup is governed by the session expiry property, which defaults to 0. This means that if you want to clean up at the start of the session, but be able to reconnect and continue if there are communication breaks until your work is done, you can do it in one go:

connect clean_start=true, session_expiry=99999 (value in seconds)

disconnect also has a session expiry property defaulting to 0, so when you issue a normal disconnect, then the session is cleaned up immediately.

So I think we could do with some setting which means "use clean start on the first connect, but not on subsequent re-connects".  This *could* be implied from the session expiry value and session present connack flag, although that might be trying to be too clever?

ralight commented:
@icraggs I agree, the current reconnection process of using the same flags as set at the initial connection is limiting. I also agree that implying it  from the session expiry value would be being too clever. I don't know what the best way of presenting it though, perhaps a further argument to the connect() call?

icraggs commented:
It might be that the natural default should be that clean_start=true only applies to the first connect and not to subsequent reconnects, as that is what I *think* most people would want most of the time.  Then the other option could be clean_start_always or similar, defaulting to false, and not allowed together with clean_start.  Or clean_start could take the possible values [True, False, FIRST_ONLY] in some way, defaulting to FIRST_ONLY.

I think that these should be on the connect call rather than the constructor, as the connect properties are on the connect also (as I have implemented it).   The only downside as far as I see it is that this could be a bit confusing for people migrating from V3.

ralight commented:
I'd support a three value option, and think it would be better in the `connect()` call. `clean_session` was in the constructor because it applied to the whole session, that's not necessarily true any more.

icraggs commented:
I've added the three value clean_start option, and put it on connect rather than the constructor.  I've also added documentation for the MQTT v5.0 functionality.  So I think that's all done, apart from the question of making the V5 tests run with tox only.

icraggs commented:
@PierreF I've moved the test broker startup to tox.ini, so the tests work when you use tox outside of Travis too.  The only situation that won't work is if one of the ports the test broker (like 1883) is already in use, probably by another MQTT broker like Mosquitto.

So that's all done from my point of view.

fpagliughi commented:
Would it be possible to maybe accept this PR into a new 'mqttv5' branch (or something like that)?

Doing so might make it easier for other people to find, test, comment upon, or possibly fix and enhance.

PierreF commented:
@icraggs as discussed by email, the last point I wanted to improve was to make easier to have callbacks that works for both MQTTv3 / MQTTv5.

The changes were rather minor, so I pushed commit on your branch. I hope you don't mind.

The changes are indented to make on_connect/on_subscribe easier to works with both MQTTv3 and MQTTv5:
* rc (and interger) / reasonCode (a object) could be compared to interger. So rc == 0 and reasonCode == 0 is valid
* Suggested to add properties=None as last parameter. So the callback could be called without properties (MQTTv3) or with properties (MQTTv5)
* Change to always provide a list of reasonCodes in on_subscribe. Since reasonCode are comparable to integer, it will behave like granted_qos (reasonCodes == [0] will works)

If my changes looks good to you, I'm going to merge it into master.

PierreF commented:
Added "six" in test dependency... we use it since addition of pytest IIRC. I think some 3th party dependency (I think to pylama) required six but now only require it with Python below 3.5.

icraggs commented:
@PierreF the changes look good to me.  Thanks!

icraggs commented:
Thanks @PierreF !
